src/main/frontend/pipeline-console-view/pipeline-console/main/hooks/use-steps-poller.ts

```diff
@@ -30,14 +30,34 @@ export function useStepsPoller(props: RunPollerProps) {
 
   const updateStepConsoleOffset = useCallback(
     async (stepId: string, forceUpdate: boolean, startByte: number) => {
-      const stepBuffer = stepBuffers.get(stepId) ?? {
-        lines: [],
-        startByte: 0 - LOG_FETCH_SIZE,
-        endByte: -1,
-        stepId,
-      };
+      let stepBuffer = stepBuffers.get(stepId);
+      if (!stepBuffer) {
+        stepBuffer = {
+          lines: [],
+          startByte: 0 - LOG_FETCH_SIZE,
+          endByte: -1,
+        };
+        stepBuffers.set(stepId, stepBuffer);
+      }
+      while (stepBuffer.pending) {
+        const pending = stepBuffer.pending;
+        const response = await pending.promise;
+        if (startByte === pending.startByte && response) {
+          return; // deduplicated fetch operation
+        }
+      }
+      if (stepBuffer.fullyFetched) return; // Already fetched in full.
       if (stepBuffer.startByte > 0 && !forceUpdate) return;
-      const response = await getConsoleTextOffset(stepId, startByte);
+      stepBuffer.pending = {
+        startByte,
+        promise: getConsoleTextOffset(stepId, startByte),
+      };
+      let response;
+      try {
+        response = await stepBuffer.pending.promise;
+      } finally {
+        delete stepBuffer.pending;
+      }
```

Comment:
I'm not sure if this intermediary assignment to `pending` is useful, could it just be
```suggestion
      let response;
      try {
        response = await getConsoleTextOffset(stepId, startByte);
      } finally {
        delete stepBuffer.pending;
      }
```

Reply:
We need the intermediate assignment for other concurrent fetch operations to observe this pending fetch operation. `stepBuffer` is a shared object in the `stepBuffers` map.

- time 00:00 fetch operation 1 starts and finds no pending operation, it populates the `pending` variable
- time 00:01 fetch operation 2 starts and finds the pending operation
- time 00:02 fetch operation 1 finishes and clears the pending operation
- time 00:02 fetch operation 2 returns early
- time 00:10 fetch operation 3 starts and finds no pending operation ... etc

Reply:
Ahh, makes sense, thanks for the explanation